cloudinit/sources/DataSourceCloudCIX.py

```diff
@@ -0,0 +1,206 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+import json
+import logging
+from typing import Optional
+
+from cloudinit import dmi, net, sources, url_helper, util
+from cloudinit.net.dhcp import NoDHCPLeaseError
+from cloudinit.net.ephemeral import EphemeralIPNetwork
+
+LOG = logging.getLogger(__name__)
+
+METADATA_URLS = ["http://169.254.169.254"]
+METADATA_VERSION = 1
+
+BUILTIN_DS_CONFIG = {
+    "retries": 3,
+    "timeout": 5,
+    "wait": 5,
+}
+
+
+CLOUDCIX_DMI_NAME = "CloudCIX"
+
+
+class DataSourceCloudCIX(sources.DataSource):
+
+    dsname = "CloudCIX"
+
+    def __init__(self, sys_cfg, distro, paths):
+        super(DataSourceCloudCIX, self).__init__(sys_cfg, distro, paths)
+        self.distro = distro
+        self.ds_cfg = util.mergemanydict(
+            [
+                util.get_cfg_by_path(sys_cfg, ["datasource", "CloudCIX"], {}),
+                BUILTIN_DS_CONFIG,
+            ]
+        )
+        self._metadata_url = None
+        self._net_cfg = None
+        self.url_retries = self.ds_cfg.get("retries")
+        self.url_timeout = self.ds_cfg.get("timeout")
+        self.url_sec_between_retries = self.ds_cfg.get("wait")
+
+    def _get_data(self):
+        """
+        Fetch the user data and the metadata
+        """
+        try:
+            with EphemeralIPNetwork(
+                self.distro,
+                interface=net.find_fallback_nic(),
+                ipv6=True,
+                ipv4=True,
+            ):
+                crawled_data = util.log_time(
+                    logfunc=LOG.debug,
+                    msg="Crawl of metadata service",
+                    func=self.crawl_metadata_service,
+                )
+        except NoDHCPLeaseError as e:
+            LOG.error("Bailing, DHCP exception: %s", e)
+            return False
+        except sources.InvalidMetaDataException as error:
+            LOG.debug(
+                "Failed to read data from CloudCIX datasource: %s", error
+            )
+            return False
+
+        self.metadata = crawled_data["meta-data"]
+        self.userdata_raw = util.decode_binary(crawled_data["user-data"])
```

Comment:
Hasn't this already been decoded in read_metadata via maybe_b64decode?

Reply:
Not quite, no. `decode_binary` turns the userdata into a Python string. It will interpret the bytes it processes as UTF-8 which will yield gibberish if the userdata happens to be Base64 encoded on the outside.